diff -burN POE-0.07-orig/MANIFEST POE-0.07/MANIFEST
--- POE-0.07-orig/MANIFEST	Tue Jun 15 11:12:58 1999
+++ POE-0.07/MANIFEST	Wed Aug 11 14:11:48 1999
@@ -45,3 +45,4 @@
 samples/thrash.perl
 samples/tutorial-chat.perl
 samples/wheels.perl
+samples/ids.perl
\ No newline at end of file
diff -burN POE-0.07-orig/POE/Kernel.pm POE-0.07/POE/Kernel.pm
--- POE-0.07-orig/POE/Kernel.pm	Fri Jun 18 14:13:01 1999
+++ POE-0.07/POE/Kernel.pm	Wed Aug 11 14:18:13 1999
@@ -110,6 +110,8 @@
 sub KR_SIGNALS        () { 4 }
 sub KR_ALIASES        () { 5 }
 sub KR_ACTIVE_SESSION () { 6 }
+sub KR_ID             () { 7 }          # -PG
+sub KR_IDs            () { 8 }          # -PG
                                         # handle structure
 sub HND_HANDLE   () { 0 }
 sub HND_REFCOUNT () { 1 }
@@ -257,6 +259,8 @@
     $self->[KR_STATES  ] = [ ];
     $self->[KR_SIGNALS ] = { };
     $self->[KR_ALIASES ] = { };
+    $self->[KR_ID      ] = '';          # -PG
+    $self->[KR_IDs     ] = { };         # -PG
                                         # initialize the vectors *as* vectors
     vec($self->[KR_VECTORS]->[VEC_RD], 0, 1) = 0;
     vec($self->[KR_VECTORS]->[VEC_WR], 0, 1) = 0;
@@ -758,6 +762,7 @@
     die "session $session already exists\a"
       if (exists $self->[KR_SESSIONS]->{$session});
   }
+  $session->ID($self->ID_alloc($session));    # -PG
 
   $self->_dispatch_state($session, $kr_active_session, EN_START, \@args);
   $self->_enqueue_state($session, $kr_active_session, EN_GC, time(), []);
@@ -809,6 +814,7 @@
 
     unless ($ss->[SS_REFCOUNT]) {
       $self->session_free($session);
+      $self->ID_free($ss->[SS_SESSION]);
     }
   }
 }
@@ -1174,6 +1180,11 @@
   if (exists $self->[KR_SESSIONS]->{$name}) {
     return $self->[KR_SESSIONS]->{$name}->[SS_SESSION];
   }
+ 
+                                        # resolve against session IDs
+  if (my $s=$self->ID_lookup($name)) { 
+    return $s if $s->isa('POE::Session');
+  }
                                         # it doesn't resolve to anything?
   $! = ESRCH;
   return undef;
@@ -1195,6 +1206,63 @@
                                         # no such session
   $! = ESRCH;
   return 0;
+}
+
+#==============================================================================
+# Kernel and Session IDs -PG
+#==============================================================================
+
+sub _ID ()
+{
+    # Attempt to get hostname
+    local $SIG{__DIE__};
+    my $host='';
+    eval {                      # POSIX::uname is fast
+        require POSIX;
+        $host=(POSIX::uname)[1];
+    } 
+#    || eval {                   # Sys::Hostname is potentialy very slow
+#        require Sys::Hostname;
+#        $host=Sys::Hostname::hostname;
+#    }
+    ;
+    $host.='-' if $host;
+    # this is guarantied to be unique localy until $$ is able
+    # to wrap around w/in one second
+    return $host.unpack 'H*', pack'N*', time, $$;
+}
+
+sub ID
+{
+    my ($self) = @_;
+    $self->[KR_ID] ||= _ID;
+    return $self->[KR_ID];
+}
+
+{
+    my $session_id=1;
+    sub ID_alloc
+    {
+        my($self, $what)=@_;
+        my $id=$session_id++;
+        # we don't use an array, because it will be sparse 
+        # (and therefore waste memory)
+        $self->[KR_IDs]->{$id}=$what;        # Not needed (for now)
+        return $id;
+    }
+}
+
+sub ID_lookup
+{
+    my($self, $id)=@_;
+    $id=$id->ID if ref $id;
+    return $self->[KR_IDs]->{$id}
+}
+sub ID_free
+{
+    my($self, $id)=@_;
+    $id=$id->ID if ref $id;    
+    delete $self->[KR_IDs]->{$id}
 }
 
 ###############################################################################
diff -burN POE-0.07-orig/POE/Session.pm POE-0.07/POE/Session.pm
--- POE-0.07-orig/POE/Session.pm	Fri Jun 18 14:13:01 1999
+++ POE-0.07/POE/Session.pm	Wed Aug 11 14:19:59 1999
@@ -28,6 +28,7 @@
 sub ARG7    () { 13 }
 sub ARG8    () { 14 }
 sub ARG9    () { 15 }
+sub _ID      () { 16 } # -PG
 
 sub SE_NAMESPACE () { 0 }
 sub SE_OPTIONS   () { 1 }
@@ -329,6 +330,18 @@
     return \%return_values;
   }
 }
+
+#------------------------------------------------------------------------------
+sub ID
+{
+    my($self, $id)=@_;
+    if(2==@_) {
+        $self->[_ID] = $id;
+        return
+    }
+    return $self->[_ID];
+}
+
 
 ###############################################################################
 1;
diff -burN POE-0.07-orig/samples/ids.perl POE-0.07/samples/ids.perl
--- POE-0.07-orig/samples/ids.perl	Wed Dec 31 19:00:00 1969
+++ POE-0.07/samples/ids.perl	Wed Aug 11 14:20:34 1999
@@ -0,0 +1,44 @@
+#!/usr/bin/perl -w
+# $Id$
+
+use lib qw(..);
+use strict;
+
+# This program tests the kernel and session IDs
+
+use POE qw(Session);
+
+sub create_session
+{
+    my($max)=@_;
+    POE::Session->new
+    (
+        _start=>sub 
+        {
+            print "Session ", $_[SESSION]->ID, " _start\n";
+            $_[KERNEL]->yield('spawn');
+        },
+        spawn=>sub
+        {
+            my $n=int rand(5);
+            print "Session ", $_[SESSION]->ID, " waiting $n seconds\n";
+            
+            $_[KERNEL]->delay("fooie$max", $n);
+        },
+        _stop=>sub
+        {
+            print "Session ", $_[SESSION]->ID, " _stop\n";
+        }
+    );
+}
+
+foreach (0..9)
+{
+    create_session($_);
+}
+
+print "POE->run()\n";
+print "Kernel ID is ", $poe_kernel->ID(), "\n";
+$poe_kernel->run();
+print "Kernel ID is still ", $poe_kernel->ID(), "\n";
+print "Done.\n";
